File: SearchInInfiniteArray.java
/*
## Problem2
Search in Infinite sorted array: 
https://leetcode.com/problems/search-in-a-sorted-array-of-unknown-size/
Given a sorted array of unknown length and a number to search for, return the index of the number in the array. Accessing an element out of bounds throws exception. If the number occurs multiple times, return the index of any occurrence. If it isn’t present, return -1.

Time complexity: O(log(n))
Space complexity: O(1) No auxilary space used to compute index
*/

class Solution {
    public int search(ArrayReader reader, int target) {
        int left = 0, right = 1, ans = -1;
        while(left <= right){
            int mid = left + (right - left)/2;
            if(reader.get(mid) == target) ans = mid;
            if(reader.get(mid) < target) {
                left = mid + 1;
                right *= 2;
            }
            else right = mid - 1;
        }
        return ans;
    }
}

File: search-in-infinite-array.py
# time complexity: O(log(m)+log(n)).
# log(m) to find the range between high & low, and log (n) to actually do the binary search
# space complexity: O(1) as no additional space used
# # Did this code successfully run on Leetcode : N/A , needed LC premium
# Any problem you faced while coding this : No

class Solution:
    def search(self, reader, target):
      # sliding window approach. If target is greater than value at high, set low = high & keep doubling high idx until your element at high is greater than target. This is the search range for your BS.
        low, high = 0, 1
        # find search range
        while reader.get(high) < target:
            low = high
            high = 2*high

        # normal binary search once you have the range
        while low <= high:
            mid = low+(high-low)//2
            if reader.get(mid) == target:
                return mid
            elif reader.get(mid) > target:
                high = mid-1
            else:
                low = mid+1
        return -1
